Externally commit regular utxo

v0d1ch opened the pull request:
This PR adds the option for our users to commit a regular utxo to the Head externally. Note that this is the initial chunk of work and we are still working on enabling script utxos to be spent in the Head too.

## Why
We want to allow our users to commit externally.

## What
Create a HTTP endpoint that can accept utxo user wants to commit to a head. Respond with a draft commit transaction that needs to be signed and submitted to the network by the user.

:palm_tree:  Added `DraftCommitTxRequest` to obtain client utxo in our server

:palm_tree: Added `DraftCommitTxResponse` to return draft tx to the user

:palm_tree: Added `APIRestInputReceived` to the logs so we can observe client contents sent 

:palm_tree: Added new `PostTxError` called `FailedToDraftTxNotInitializing` in case we experience a failure while drafting a tx


---

<!-- Consider each and tick it off one way or the other -->
* [x] CHANGELOG updated or not needed
* [x] Documentation updated or not needed
* [x] Haddocks updated or not needed
* [x] No new TODOs introduced or explained herafter

ffakenz committed:
Add draft test for building drafted commit tx

v0d1ch committed:
Add initial implementation of draftCommitTx

ffakenz committed:
Add base e2e draft for external ommit

Ignore previously added hydra-node spec.

v0d1ch committed:
Introduce the http call to get Draft commit tx

v0d1ch committed:
Start work on the http response

ffakenz committed:
Fix http server http method and client call from e2e

ffakenz committed:
Add APIRestInputReceived server log

v0d1ch committed:
Introduce rest api callback

v0d1ch committed:
Move rest callback to Server

v0d1ch committed:
Introduce Chain handle to draft commit tx

v0d1ch committed:
Sign the commit tx and introduce external key

v0d1ch committed:
Balance the commit transaction

v0d1ch committed:
Checkpoint - missing script input after balancing

v0d1ch committed:
Use commit instead of commitTx

v0d1ch committed:
Dirty solution that works

v0d1ch committed:
Remove all warnings

v0d1ch committed:
Refactor draftTx

v0d1ch committed:
Remove unneeded todo

v0d1ch committed:
Remove MonadCatch constraint

ffakenz committed:
Remove spec not needed

ffakenz committed:
Minor fix error and warnings

v0d1ch committed:
Move http server types to appropriate place

v0d1ch committed:
Add new server inputs/outputs to api.yaml

v0d1ch committed:
Alter to/from json instances for RestServerOutput

v0d1ch committed:
Add changelog entry

v0d1ch committed:
Change wording in draftTx dummy handle

v0d1ch committed:
Improve on api.yaml wording

v0d1ch committed:
Revert the State module

- Changes here are not needed after all

v0d1ch committed:
Small refactor and haddocks

v0d1ch committed:
Include new keys in data-dir

v0d1ch committed:
Improve on json schema

ch1bo commented on hydra-cluster/config/credentials/external.sk:
As we don't pass these to any programs, we should not need dedicated files for them. Then, also no need to create an "Actor" for this.

ch1bo commented on hydra-cluster/src/Hydra/Cluster/Scenarios.hs:
```diff
@@ -135,6 +143,66 @@ singlePartyHeadFullLifeCycle tracer workDir node@RunningNode{networkId} hydraScr
+  traceRemainingFunds actor = do
+    (actorVk, _) <- keysFor actor
+    (fuelUTxO, otherUTxO) <- queryMarkedUTxO node actorVk
+    traceWith tracer RemainingFunds{actor = actorName actor, fuelUTxO, otherUTxO}
```
DRY or not do this.

ch1bo commented on hydra-cluster/src/Hydra/Cluster/Scenarios.hs:
```diff
@@ -135,6 +143,66 @@ singlePartyHeadFullLifeCycle tracer workDir node@RunningNode{networkId} hydraScr
+    aliceChainConfig <-
+      chainConfigFor Alice workDir nodeSocket [] contestationPeriod
+        <&> \config -> config{networkId, startChainFrom = Just tip}
+    let hydraNodeId = 1
```
No need to start from tip, this can be simplified.

ch1bo commented on hydra-cluster/src/Hydra/Cluster/Scenarios.hs:
```diff
@@ -5,26 +5,34 @@ module Hydra.Cluster.Scenarios where
 import Hydra.Options (ChainConfig, networkId, startChainFrom)
 import Hydra.Party (Party)
 import HydraNode (EndToEndLog (..), input, output, send, waitFor, waitForAllMatch, waitMatch, withHydraNode)
+import Network.HTTP.Req
```
Should: use explicit imports

ch1bo commented on hydra-cluster/hydra-cluster.cabal:
```diff
@@ -125,7 +127,9 @@ library
     , cardano-ledger-core
+    , cardano-ledger-shelley
     , cardano-slotting
+    , cborg
```
These two imports are a bit surprising? Why do we need them?

v0d1ch replied:
we don't

ch1bo commented on hydra-node/json-schemas/api.yaml:
```diff
@@ -82,6 +84,7 @@ channels:
           - $ref: "#/components/messages/InvalidInput"
           - $ref: "#/components/messages/PostTxOnChainFailed"
           - $ref: "#/components/messages/CommandFailed"
+          - $ref: "#/components/messages/RestServerOutput"
```
This is not sent over the websocket, remove it.

ch1bo replied:
We expect another channel/endpoint for this instead.

ch1bo commented on hydra-node/json-schemas/api.yaml:
```diff
@@ -96,6 +99,7 @@ channels:
           - $ref: "#/components/messages/Close"
           - $ref: "#/components/messages/Contest"
           - $ref: "#/components/messages/Fanout"
+          - $ref: "#/components/messages/RestClientInput"
```
This is not accepted over the websocket, remove it.

ch1bo commented on hydra-node/json-schemas/api.yaml:
```diff
@@ -700,6 +720,22 @@ components:
+            type: string
+            enum: ["RestServerOutput"]
+          commitTx:
+            $ref: "#/components/schemas/RawTransaction"
```
Should: have a test checking this schema against our encoders/decoders.

Maybe something similar like [here](https://github.com/input-output-hk/hydra/blob/06078074cc72ed01f10de56c840da88e6be7ec06/hydra-node/test/Hydra/LoggingSpec.hs#L32-L33) or [here](https://github.com/input-output-hk/hydra/blob/06078074cc72ed01f10de56c840da88e6be7ec06/docs/package.json#L10-L12)

ch1bo commented on hydra-node/src/Hydra/API/ServerOutput.hs:
Move Rest request and response types into their own module (one or two). They have nothing to do with the client inputs and server outputs.

ch1bo commented on hydra-node/src/Hydra/API/ServerOutput.hs:
```diff
@@ -148,6 +149,44 @@ data ServerOutputConfig = ServerOutputConfig
   }
   deriving (Eq, Show)
 
+newtype RestServerOutput tx = DraftedCommitTx
```
Should: not indicate there will be more sum type constructors. YAGNI and might never make sense.

This is is basically a `DraftCommitTxResponse` if you want to be very specific.

ch1bo commented on hydra-node/src/Hydra/API/ServerOutput.hs:
```diff
@@ -148,6 +149,44 @@ data ServerOutputConfig = ServerOutputConfig
+    Aeson.object
+      [ "tag" Aeson..= Aeson.String "RestServerOutput"
+      , "commitTx" Aeson..= (String . decodeUtf8 . Base16.encode $ serialize' tx)
+      ]
```
Should: add a roundtrip (and golden) test whenever we hand-write decoder/encoders

ch1bo commented on hydra-node/src/Hydra/API/ServerOutput.hs:
```diff
@@ -148,6 +149,44 @@ data ServerOutputConfig = ServerOutputConfig
+  toJSON (DraftedCommitTx tx) =
+    Aeson.object
+      [ "tag" Aeson..= Aeson.String "RestServerOutput"
+      , "commitTx" Aeson..= (String . decodeUtf8 . Base16.encode $ serialize' tx)
```
Explicit import .= and operators in general (for better readability)

ch1bo commented on hydra-node/src/Hydra/API/Server.hs:
```diff
@@ -115,7 +126,7 @@ withAPIServer host port party PersistenceIncremental{loadAll, append} tracer cal
   (notifyServerRunning, waitForServerRunning) <- setupServerNotification
   race_
-    (runAPIServer host port party tracer history callback headStatusP snapshotUtxoP responseChannel notifyServerRunning)
+    (runAPIServer host port party tracer history router callback headStatusP snapshotUtxoP responseChannel notifyServerRunning)
```
I don't understand why this would be called `router`. I would expect `chain` for ting of type `Chain`

ch1bo commented on hydra-node/src/Hydra/API/Server.hs:
```diff
@@ -302,6 +324,28 @@ runAPIServer host port party tracer history callback headStatusP snapshotUtxoP r
 
+  -- Handle user requests to obtain a draft commit tx
+  handleDraftCommitUtxo directChain req respond = do
+    body <- getWaiRequestBody req
```
Use `consumeRequestBodyStrict` instead

ch1bo replied:
Or just use lazy IO, by doing `Aeson.decode <$> consumeRequestBodyLazy req`

pgrange reviewed (changes requested):
I see the following issues with this feature:
1. the hydra-node is paying the fees for the client, is it ok in the long run?
2. ~~the hydra-node could spend the same UTxO twice for paying fees if it receives two draftTx in a row. Is it ok ?~~
3. the client could ask the hydra-node to spend a UTxO that is owned by the node and not the client. Is it ok in the long run?

I don't think those 3 points are ok. I would need to ensure that we keep track of them for future evolutions.

ch1bo commented on hydra-node/src/Hydra/API/Server.hs:
```diff
@@ -302,6 +324,28 @@ runAPIServer host port party tracer history callback headStatusP snapshotUtxoP r
+            { method = decodeUtf8 $ requestMethod req
+            , paths = pathInfo req
+            , requestInputBody = Just $ toJSON requestInput
+            }
```
Could: log requests into `tracer` instead using some middleware, e.g. https://hackage.haskell.org/package/wai-extra-3.1.13.0/docs/Network-Wai-Middleware-RequestLogger.html#t:Callback

ch1bo commented on hydra-node/src/Hydra/API/Server.hs:
```diff
@@ -1,4 +1,5 @@
 {-# LANGUAGE QuasiQuotes #-}
+{-# LANGUAGE TupleSections #-}
```
Where is this needed?

ch1bo commented on hydra-node/src/Hydra/Chain.hs:
```diff
@@ -184,6 +186,7 @@ newtype Chain tx m = Chain
   -- reasonable local view of the chain and throw an exception when invalid.
   --
   -- Does at least throw 'PostTxError'.
+  , draftTx :: (IsChainState tx, MonadThrow m) => UTxOType tx -> m (Either (PostTxError tx) tx)
```
Should: document this new method and in particular what types of exceptions are thrown.

ch1bo replied:
@v0d1ch this is still missing haddock documentation

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Handlers.hs:
```diff
@@ -144,8 +145,14 @@ mkChain tracer queryTimeHandle wallet ctx LocalChainState{getLatest} submitTx =
+        chainState <- atomically getLatest
+        case Hydra.Chain.Direct.State.chainState chainState of
+          Initial st ->
+            finalizeTx wallet ctx chainState utxo `traverse` commit ctx st utxo
```
Maybe we should limit not allow to commit from the internal wallet owned UTxO, or maybe this is exactly what we want to replace the "internal" commit by this "external" commit.

ch1bo commented on hydra-node/src/Hydra/Ledger/Cardano.hs:
```diff
@@ -261,6 +261,15 @@ genOutput vk = do
+  pure $ TxOut (mkVkAddress (Testnet $ NetworkMagic 42) vk) value TxOutDatumNone ReferenceScriptNone
+
 -- | Generate an ada-only 'TxOut' payed to an arbitrary public key.
 genTxOutAdaOnly :: Gen (TxOut ctx)
```
We only use `genTxOutAdaOnly` in two places, so you could just parameterize it by `VerificationKey` instead of introducing another function.

ch1bo reviewed (commented):
Comments from our discussion today.

v0d1ch committed:
Remove external key pair

v0d1ch committed:
Remove unneeded deps

v0d1ch committed:
Use explicit imports in Scenarios.hs

v0d1ch committed:
Add json roundtrip tests for new http types

v0d1ch committed:
Extract request handler

- Remove not needed pragma
- Use consumeRequestBodyStrict
- Rename router to chain

v0d1ch committed:
Extract rest types into separate module

v0d1ch committed:
Rename new http server types

v0d1ch committed:
Add golden tests for new types

v0d1ch committed:
Rename FailedToDraftTx to FailedToDraftTxNotInitializing

- Decide on what errors we should throw in a Chain handle
- Improve haddocks

v0d1ch committed:
Remove genAdaOnlyOutput and make genTxOutAdaOnly accept a public key

v0d1ch commented:
@pgrange I think your comments were addressed in the grooming session.

v0d1ch committed:
Remove removed constructor

ch1bo commented on hydra-node/json-schemas/api.yaml:
```diff
@@ -3,15 +3,17 @@ info:
   title: Hydra Node API
   version: '0.10.0'
```
Should: update to 0.11.0

ch1bo commented on hydra-node/json-schemas/api.yaml:
```diff
@@ -97,6 +111,23 @@ channels:
+
+    publish:
+      summary: Requests sent to the Hydra node.
+      operationId: restClientInput
```
Should: update the operation ids here and on line 120 to somehow be specific to the draft tx endpoint.

ffakenz replied:
good catch!

ch1bo commented on hydra-node/json-schemas/api.yaml:
```diff
@@ -97,6 +111,23 @@ channels:
+          - $ref: "#/components/messages/DraftCommitTxResponse"
+
+    publish:
+      summary: Requests sent to the Hydra node.
```
Must: mention the HTTP verb to be used

There are HTTP bindings which can be used on channel or message level to describe this, but documenting this in the summary / description would also be fine.

ch1bo commented on hydra-node/json-schemas/api.yaml:
```diff
@@ -218,6 +249,22 @@ components:
+        type: object
+        required:
+          - tag
+          - utxo
```
Could: rename this to utxos or describe above that zero, one or more outputs can be committed. Alternatively, add a description to `UTxO` schema explaining this is actually a map of transaction outrefs to unspent transaction outputs.

ch1bo commented on hydra-node/src/Hydra/API/ClientInput.hs:
```diff
@@ -1,3 +1,4 @@
+{-# LANGUAGE DuplicateRecordFields #-}
```
Should: remove likely redundant pragma

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Handlers.hs:
```diff
@@ -144,8 +141,23 @@ mkChain tracer queryTimeHandle wallet ctx LocalChainState{getLatest} submitTx =
+    , -- Handle that creates a draft commit tx using the user utxo. Here we
+      -- distinguish between errors users can do something about (eg.
+      -- 'CannotCommitReferenceScript' which are returned in Left ) and the
+      -- system errors that will show as 500 errors on the api server level.
```
I think you would want to move this explanation to the `draftTx` definition.

Also, speaking about "users" and "500 errors" are out of context here.

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Handlers.hs:
```diff
@@ -144,8 +141,23 @@ mkChain tracer queryTimeHandle wallet ctx LocalChainState{getLatest} submitTx =
+            case commit ctx st utxo of
+              Left CannotCommitReferenceScript -> pure $ Left CannotCommitReferenceScript
+              Left (CommittedTooMuchADAForMainnet l ml) -> pure $ Left $ CommittedTooMuchADAForMainnet l ml
+              Left e -> throwIO e
```
Must: make this consistent as this is missing at least the `UnsupportedLegacyOutput`

It's weird to take the decision here on "what can be dealt with by the user" and what not. i.e. the separation between `Left` and `throwIO`.

If we want to stick to "checked" or enumerated errors, then we should have a dedicated type for possible errors coming from the `commit` function (should be 4), which is to be wrapped into yet another type for the additional `FailedToDraftTxNotInitializing` case here. Then, only at the call-site of `draftTx` we can take a decision, which of the then 5 possible errors are bad requests and which are internal server errors.

Or we keep the (very) big `PostTxError` here and remove the `MonadThrow m` constraint.. underlining that `draftTx` can only produce a `Left` with any of these cases. On the call site we need to pick then out the bad requests and treat everything else as an internal error.

v0d1ch replied:
I moved the error handling at the call site but we need `MonadThrow` constraint because of `finalizeTx` usage.

ch1bo commented on hydra-node/test/Hydra/API/RestServerSpec.hs:
```diff
@@ -0,0 +1,40 @@
+settings =
+  defaultSettings
+    { sampleSize = 200
+    }
```
Must: not bloat golden files

200 is way to much for simple product types `DraftCommitTxRequest` and `DraftCommitTxResponse`. The default (5) should suffice.

Also, I think we don't need `ReasonablySized` here and maybe only one transaction type (I'd pick the cardano `Tx`) is enough.

ch1bo replied:
Okay we need `ReasonablySized` for the request type. As you can see, the golden file is huge while not adding much variability on the request data.

ch1bo commented on hydra-node/test/Hydra/API/ServerSpec.hs:
```diff
@@ -382,6 +390,9 @@ testClient queue semaphore cnx = do
       testClient queue semaphore cnx
 
+dummyChainHandle :: Chain tx IO
+dummyChainHandle = Chain{postTx = \_ -> pure (), draftTx = \_ -> error "draftTx not implemented"}
```
Must: use `error` on `postTx` as we expect none of these tests to call this function.

Also, draftTx is implemented, but just not expected to be invoked here.
Also, as this `dummyChainHandle` is used often here, you _could_ DRY the invocation using a local definition using this mocked handle `withTestAPIServer`.

ch1bo commented on hydra-node/test/Hydra/BehaviorSpec.hs:
```diff
@@ -594,7 +594,7 @@ simulatedChainAndNetwork initialChainState = do
           pure $
             node
-              { oc = Chain{postTx = postTx nodes history chainStateVar}
+              { oc = Chain{postTx = postTx nodes history chainStateVar, draftTx = \_ -> error "draftTx not implemented"}
```
Could: fail with "unexpected call to draftTx" to be clear it's not expected

ch1bo commented on hydra-node/test/Hydra/LoggingSpec.hs:
```diff
@@ -31,6 +32,8 @@ spec = do
             [ prop_validateToJSON @(Envelope (HydraLog Tx ())) (dir </> "logs.yaml") "messages" (dir </> "HydraLog")
             , prop_specIsComplete @(HydraLog Tx ()) (dir </> "logs.yaml") apiSpecificationSelector
+            , prop_specIsComplete @(DraftCommitTxResponse Tx) (dir </> "logs.yaml") apiSpecificationSelector
+            , prop_specIsComplete @(DraftCommitTxRequest Tx) (dir </> "logs.yaml") apiSpecificationSelector
```
This should not pass!?

`DraftCommitTxRequest` is not specified in `logs.yaml`, and certainly not at path `key "properties" . key "message"`.

ffakenz commented on hydra-cluster/src/Hydra/Cluster/Scenarios.hs:
```diff
@@ -135,6 +143,55 @@ singlePartyHeadFullLifeCycle tracer workDir node@RunningNode{networkId} hydraScr
+    withHydraNode tracer aliceChainConfig workDir hydraNodeId aliceSk [] [1] hydraScriptsTxId $ \n1 -> do
+      -- Initialize & open head
+      send n1 $ input "Init" []
+      headId <- waitMatch 600 n1 $ headIsInitializingWith (Set.fromList [alice])
```
maybe 600 is too much?

v0d1ch replied:
Reduced it to 60 instead

ffakenz commented on hydra-cluster/src/Hydra/Cluster/Scenarios.hs:
```diff
@@ -135,6 +143,55 @@ singlePartyHeadFullLifeCycle tracer workDir node@RunningNode{networkId} hydraScr
+      let signedCommitTx = signWith externalSk commitTx
+      submitTransaction networkId nodeSocket signedCommitTx
+
+      waitFor tracer 600 [n1] $
```
same

ffakenz committed:
Improve api yaml for rest api

- update api version to 0.11.0
- update  api description and operation ids
- upgrade  npm package to fix yarn run v1.22.19
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command. cmd
- add http bindings to api /commit channel
- add tag to response to make yarn run v1.22.19
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command. happy, but it is still broken for request

v0d1ch committed:
Rename utxo to utxos in DraftCommitTxRequest

v0d1ch committed:
Generate default samples without ReasonablySized files

v0d1ch committed:
Handle errors at the call site of draftTx

v0d1ch committed:
Reduce wait time to 60

v0d1ch committed:
Add withTestAPIServer

v0d1ch committed:
Remove uneeded assertion in LoggingSpec

v0d1ch committed:
Do not handle CannotFindOwnInitial at the api level

ch1bo commented on hydra-node/json-schemas/api.yaml:
```diff
@@ -97,6 +111,52 @@ channels:
+              tag:
+                type: string
+              utxos:
+                $ref: "#/components/schemas/UTxO"
```
Must: not use `query`

This is for query parameters and as far as I can tell we are not considering any query params, but just have a request body.

ch1bo commented on hydra-node/json-schemas/api.yaml:
```diff
@@ -97,6 +111,52 @@ channels:
+        http:
+          type: response
+          method: POST
+          query:
```
Must: not use `query`

The response does use a query parameters, but only a response body.

ch1bo commented on hydra-node/json-schemas/api.yaml:
```diff
@@ -97,6 +111,52 @@ channels:
+      summary: Hydra node drafted commit transaction responses.
+      operationId: draftCommitTxResponse
+      message:
+        $ref: "#/components/messages/DraftCommitTxResponse"
```
Should: Document the possible error responses with status codes.

Could: inline response and error schemas as they likely are not going to be re-used

ffakenz committed:
Remove query field on http bindings as they are not needed

The query field represents a Schema object containing the definitions for each query parameter.
We are not considering any query params, but just have a request body.
The response does use a query parameters either, but only a response body.

v0d1ch committed:
Use ReasonablySized for new draft commit types

ch1bo reviewed (approved):
Good enough to continue with the other work streams
